### PR TITLE
feat(aurora-portal): implement Ceph bucket versioning BFF endpoints

### PR DESCRIPTION
**Pull request by KirylSAP**
# Summary

This PR implements S3 bucket versioning support for Ceph Object Storage, enabling users to maintain multiple variants of objects, recover from accidental deletions, and manage object history through the Aurora BFF layer.

Versioning allows keeping multiple variants of an object in the same bucket, providing protection against accidental overwrites and deletions. Once enabled, versioning cannot be fully disabled (only suspended), and each object version receives a unique `versionId`.

# Changes Made

## Core Implementation

- **tRPC Router** (`versioningRouter.ts` - 202 lines)
  - `getStatus` - get current versioning status for a bucket
  - `setStatus` - enable/suspend versioning with validation
  - `listVersions` - list all object versions with pagination
  - `listObjectVersions` - get versions for a specific object key
  - `deleteVersion` - permanently delete a version
  - `restoreVersion` - restore a previous version

- **Type Definitions** (`versioning.ts`)
  - Comprehensive TypeScript types and Zod schemas for all versioning operations
  - Input/output validation schemas
  - Type-safe API contracts

- **S3 Error Mapping** (`s3ErrorMapper.ts`)
  - Added `NoSuchVersion` → `NOT_FOUND`
  - Added `InvalidBucketState` → `BAD_REQUEST`
  - Added `VersioningNotEnabled` → `PRECONDITION_FAILED`

- **Comprehensive Tests** (`versioningRouter.test.ts` - 382 lines)
  - 100% coverage of all versioning endpoints
  - Mock S3 client for isolated testing
  - Edge cases: missing credentials, null version IDs, pagination, delete markers

## Security & Bug Fixes

- **Fix URL Injection (High severity)**
  - Apply `encodeURIComponent()` to object keys in `restoreVersion()`
  - Prevents malformed S3 requests when keys contain special characters (?, &, spaces, /)
  - Matches pattern used in `objectRouter.ts` copy operations

- **Race Condition Documentation**
  - Document that `getVersioningStatus()` provides point-in-time snapshot
  - Single S3 command ensures consistent `Status` and `MFADelete` values

- **Extended Test Coverage**
  - Tests for URL encoding with spaces (`"my folder/my file.txt"`)
  - Tests for query characters (`"file?param=value&other=data.txt"`)
  - Verify `CopySource` contains properly encoded keys

# Related Issues

- S3 Epic: Ceph Object Storage BFF Implementation
- Task 10: Implement S3 Bucket Versioning Endpoints

# Architecture Notes

## Service Layer Pattern
- **Router** → **Service** → **S3 Client**: Clean separation of concerns
- Router handles tRPC validation and auth via `cephProtectedProcedure`
- Service encapsulates S3 SDK logic with TypeScript types
- Consistent error mapping via `tryCatchWithContext()` and `mapS3ErrorToTRPCError()`

## Key Design Decisions

1. **Versioning Cannot Be Disabled**
   - Once enabled, can only be suspended (S3 limitation)
   - Service layer enforces this in `setVersioningStatus()`

2. **Restore = Copy**
   - S3 has no native "promote version" operation
   - `restoreVersion()` copies old version to create new latest version
   - All versions preserved (including old and restored)

3. **Delete Markers**
   - Deleting an object in versioned bucket creates a "delete marker"
   - `listVersions()` returns both versions and delete markers separately
   - Deleting a delete marker "undeletes" the object

4. **Pagination**
   - Uses S3's dual-marker system (`KeyMarker` + `VersionIdMarker`)
   - Max 1000 items per request (capped in validation)
   - Frontend handles pagination with `nextKeyMarker`/`nextVersionIdMarker`

# Testing Instructions

## Prerequisites
```bash
pnpm install
```

## Run Tests
```bash
# All tests (139 test files, 2942 tests)
pnpm test

# Versioning-specific tests
pnpm test -- --exclude="" src/server/Storage/routers/ceph/versioningRouter.test.ts
```

### 1. Enable Versioning
```bash
curl -X POST http://localhost:3000/polaris-bff/storage.ceph.versioning.setStatus \
  -H "Content-Type: application/json" \
  -d '{
    "project_id": "your-project-id",
    "bucket": "test-bucket",
    "status": "Enabled"
  }'
```

### 2. Check Status
```bash
curl http://localhost:3000/polaris-bff/storage.ceph.versioning.getStatus?project_id=...&bucket=test-bucket
```

### 3. List Versions
```bash
curl http://localhost:3000/polaris-bff/storage.ceph.versioning.listVersions?project_id=...&bucket=test-bucket&maxKeys=10
```

### 4. Restore Version
```bash
curl -X POST http://localhost:3000/polaris-bff/storage.ceph.versioning.restoreVersion \
  -d '{
    "project_id": "...",
    "bucket": "test-bucket",
    "key": "my-object.txt",
    "versionId": "version-abc123"
  }'
```

### 5. Delete Version (Permanent)
```bash
curl -X POST http://localhost:3000/polaris-bff/storage.ceph.versioning.deleteVersion \
  -d '{
    "project_id": "...",
    "bucket": "test-bucket",
    "key": "my-object.txt",
    "versionId": "version-abc123"
  }'
```

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes (139 test files, 2942 tests).
- [x] I have made corresponding changes to the documentation (JSDoc comments).
- [x] My changes generate no new warnings or errors.
- [x] Code reviewed by triple-agent system (security, correctness, architecture).

# Additional Notes

## Security Considerations
- All endpoints protected by `cephProtectedProcedure` (requires EC2 credentials)
- Project-scoped via `projectScopedProcedure`
- Authorization delegated to S3's native IAM
- Error messages sanitized via `mapS3ErrorToTRPCError()`

## Performance Notes
- List operations capped at 1000 items per request
- Pagination supported for large result sets
- Service layer caches nothing (stateless)

## Summary by CodeRabbit

* **New Features**
  * Added S3 bucket and object versioning support, including ability to manage versioning status, list versions, restore previous versions, and delete specific versions.

* **Documentation**
  * Updated Ceph S3 BFF documentation with comprehensive versioning procedure details and S3 error code mappings.

* **Tests**
  * Added test suite for versioning operations.

### CODE REVIEW

File: packages/aurora/src/client/trpcClient.ts [hunk 1]
```diff
@@ -91,7 +91,7 @@ const getLinks = () => [
 ]
 
 // React Query client (for hooks like useQuery/useMutation/useSubscription)
-export const trpcReact = createTRPCReact<AuroraRouter>()
+export const trpcReact: ReturnType<typeof createTRPCReact<AuroraRouter>> = createTRPCReact<AuroraRouter>()
```
[the remark this answers is not in the document]
Comment (KirylSAP) [reply packages/aurora/src/client/trpcClient.ts:94]: ReturnType<typeof createTRPCReact<AuroraRouter>> is valid TypeScript syntax

**Review by andypf — APPROVED** [review]

**Review by vlad-schur-external-sap — APPROVED** [review]
